source/MaterialXRuntime/Private/PvtApi.cpp

```diff
@@ -4,81 +4,183 @@
-    _stages.clear();
-    _nodedefs.clear();
-    _nodeimpls.clear();
-
-    _libraryRootStage.reset();
     _libraries.clear();
-    _libraryRootStage = RtStage::createNew(libRootName);
+    _librariesOrder.clear();
+    _stages.clear();
+    _stagesOrder.clear();
 
     _unitDefinitions = UnitConverterRegistry::create();
 }
 
-void PvtApi::createLibrary(const RtToken& name)
+RtStagePtr PvtApi::loadLibrary(const RtToken& name, const FilePath& path, const RtReadOptions* options, bool forceReload)
 {
-    // If already loaded unload the old first,
-    // to support reloading of updated libraries.
-    if (getLibrary(name))
+    auto it = _libraries.find(name);
+    if (it != _libraries.end())
+    {
+        if (forceReload)
+        {
+            unloadLibrary(name);
+        }
+        else
+        {
+            throw ExceptionRuntimeError("A library named '" + name.str() + "' has already been loaded");
+        }
+    }
+
+    RtStagePtr stage = PvtStage::createNew(name);
+    _libraries[name] = stage;
+    _librariesOrder.push_back(stage);
+
+    // Load in the files(s).
+    RtFileIo fileApi(stage);
+    StringSet loadedFiles = fileApi.readLibrary(path, _searchPaths, options);
+    for (const string& file : loadedFiles)
     {
-        unloadLibrary(name);
+        stage->addSourceUri(file);
     }
 
-    RtStagePtr lib = RtStage::createNew(name);
-    _libraries[name] = lib;
+    // Register any definitions and implementations.
+    registerPrims(stage);
 
-    _libraryRootStage->addReference(lib);
+    // Reset nodeimpl relationsships since the registry of
+    // definitions and implementations have changed.
+    setupNodeImplRelationships();
+
+    return stage;
 }
 
-void PvtApi::loadLibrary(const RtToken& name, const RtReadOptions& options)
+void PvtApi::unloadLibrary(const RtToken& name)
 {
-    // If already loaded unload the old first,
-    // to support reloading of updated libraries.
-    if (getLibrary(name))
+    auto it = _libraries.find(name);
+    if (it != _libraries.end())
     {
-        unloadLibrary(name);
-    }
+        RtStagePtr stage = it->second;
+        unregisterPrims(stage);
 
-    RtStagePtr lib = RtStage::createNew(name);
-    _libraries[name] = lib;
+        _libraries.erase(it);
+        _librariesOrder.erase(std::find(_librariesOrder.begin(), _librariesOrder.end(), stage));
 
-    RtFileIo file(lib);
-    file.readLibraries({ name.str() }, _searchPaths, options);
+        // Reset nodeimpl relationsships since the registry of
+        // definitions and implementations have changed.
+        setupNodeImplRelationships();
+    }
+}
 
-    _libraryRootStage->addReference(lib);
+void PvtApi::unloadLibraries()
+{
+    for (auto stage : _librariesOrder)
+    {
+        unregisterPrims(stage);
+    }
+    _libraries.clear();
+    _librariesOrder.clear();
 }
 
-void PvtApi::unloadLibrary(const RtToken& name)
+void PvtApi::registerPrims(RtStagePtr stage)
+{
+    for (RtPrim prim : stage->traverse())
```

Comment:
Seems like we should clear the previous prims before reloading here? Or we rely on duplicate checking.

Reply:
This is already handled internally in each register method. For example, see the implementation of registerNodeDef, registerNodeGraph, etc.